Instrument pypy raw

vext01 opened the pull request:
Write out raw PyPy events into results file, e.g.:

Large comment in the code explains the format.

```
 "inst_data": {
  "nbody:PyPy:default-python": {
   "raw_vm_events": [
    [
     [
      "root", 
      null, 
      null, 
      [
       [
        "gc-set-nursery-size", 
        32583494957343, 
        32583495128488, 
        []
       ], 
       [
        "gc-minor", 
        32583495215578, 
        32583495272469, 
        [
         [
          "gc-minor-walkroots", 
          32583495222035, 
          32583495227446, 
          []
         ]
        ]
       ], 
       [
        "gc-set-nursery-size", 
        32583495284669, 
        32583495294252, 
        []
       ], 
```

By all means review, but don't merge yet please. I want to add some tests and try it on a reduced version of the warmup experiment.

vext01 commented:
Er, github is confused. Some of the old PR comments are here...

ltratt commented:
Looks good to me. Let me know when you've tested, and I'll merge.

vext01 commented:
Added some tests. VMs still building on bencher2...

vext01 commented:
I've just brought this up to date with master, so I can test the changes to the iterations runners...

vext01 commented:
The iterations runners changes appear to work OK. There were some unrelated crashes to do with Oracle VM updates which I will handle in separate PRs.

OK for me to squash @ltratt?

ltratt commented on iterations_runners/iterations_runner.py:
```diff
@@ -45,6 +47,11 @@
         bench_func(param)
         stop_time = clock_gettime_monotonic()
 
+        # In instrumentation mode, write a iteration separator to stderr.
```
"an iteration separator"

ltratt commented on krun/results.py:
```diff
@@ -92,6 +100,12 @@ def write_to_file(self):
             f.write(json.dumps(to_write,
                                indent=1, sort_keys=True, encoding='utf-8'))
 
+    def add_instrumentation_data(self, bench_key, inst_dct):
```
This method name is a real mouthful... Can we do `add_instr_data` or similar?

ltratt commented:
There are large numbers of:

```python
    def parse_instrumentation_stderr_file(self, file_handle):
        pass
```

which suggests to me the superclass should have this method as the default?

ltratt commented:
OK, a few minor comments on the new commits.

vext01 commented:
The `pass` methods are implementing abstract methods. This is a design decision. I am forcing the implementer to think about what needs to be done to support a new VM.

ltratt commented:
OK, but if they're all `pass`...?

vext01 commented:
Java and PyPy will have implementations of `parse_instrumentation_stderr_file`.

ltratt commented:
I dunno, I'm not entirely convinced in this case, because if someone doesn't provide a `parse_instr_stderr` (which is a snappier name too) method, then things still work OK? In other words, it'll be pretty obvious to them if they haven't dealt with things.

Maybe a more fundamental question just raised itself to me: is this parsing happening while the iterations runner is happening? Or is it post-processing? Or some combination?

vext01 commented:
If an implementation isn't provided, things will work. Maybe it isn't a huge deal in this case...

Parsing is all post.

ltratt commented:
OK, so let's make it a `pass` method in the superclass then? [Can we shorten the name at the same time too? Long method names make code harder to read IMHO, as it causes lots of line spills.]

vext01 commented:
Way ahead of you ;)

ltratt commented:
I expected no less :)

vext01 commented:
Fixed. Caught a bug and some more items that should be `instr_...`. OK to squash?

vext01 commented:
Not sure why GH put my comment in the middle of the changeset :P

ltratt commented:
Looks good... squash and I'll merge?

vext01 commented:
done

ltratt commented:
Something just changed? I think there was a second force push.

vext01 commented:
yeah, fixed the test

vext01 commented:
I should have pushed the new commit, so you could see what changed.

There was a test exercising the bogus assertion that I fixed in the last round. I removed this.

ltratt commented:
OK, mistakes happen, no probs.
